### PR TITLE
implement gatherND

### LINKED ISSUE #87: Implement gatherNd
```spec
MLOperand gatherND(MLOperand input, MLOperand indices, optional MLGatherNDOptions options = {});
```

Refer to https://onnx.ai/onnx/operators/onnx__GatherND.html

### PR DESCRIPTION
**Pull request by BruceDai**
This PR is to fix #87.

@fdwr @huningxin @shiyi9801 PTAL, thanks!

### CODE REVIEW

**Review by fdwr — APPROVED** [review]
👍

File: src/gather_nd.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+  const inputRank = input.rank;
+  const inputShape = input.shape;
+  const indicesRank = indices.rank;
+  const indicesShape = indices.shape;
+  const lastIndicesSize = indicesShape[indicesRank - 1];
+  const tmpShape = inputShape.slice(lastIndicesSize);
+
+  // Refer to https://docs.openvino.ai/2024/documentation/openvino-ir-format/operation-sets/ +
+  //   operation-specs/movement/gather-nd-8.html
```
Comment (huningxin) [inline src/gather_nd.js:23]: Nit: please keep the URL in one line.

File: src/gather_nd.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+ */
+export function gatherND(input, indices) {
+  validateGatherNDParams(input, indices);
+
+  const inputRank = input.rank;
+  const inputShape = input.shape;
+  const indicesRank = indices.rank;
+  const indicesShape = indices.shape;
+  const lastIndicesSize = indicesShape[indicesRank - 1];
+  const tmpShape = inputShape.slice(lastIndicesSize);
```
Comment (huningxin) [inline src/gather_nd.js:20]: `sliceShape`: it represents the shape of a slice that the operator gathers from the input tensor.

```suggesting
```

```suggestion
  const sliceShape = inputShape.slice(lastIndicesSize);
```

File: src/gather_nd.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+  // Refer to https://docs.openvino.ai/2024/documentation/openvino-ir-format/operation-sets/ +
+  //   operation-specs/movement/gather-nd-8.html
+  let outputShape = indicesShape.slice(0, indicesRank - 1);
+  if (lastIndicesSize !== inputRank) {
+    outputShape = outputShape.concat(tmpShape);
+  }
+
+  const output = new Tensor(outputShape);
+  const indicesTotal = sizeOfShape(indicesShape);
```
Comment (huningxin) [inline src/gather_nd.js:30]:
```suggestion
  const indicesSize = sizeOfShape(indicesShape);
```

File: src/gather_nd.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  const tmp = new Tensor(tmpShape);
```
Comment (huningxin) [inline src/gather_nd.js:31]:
```suggestion
  const slice = new Tensor(tmpShape);
```

File: src/gather_nd.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  const tmpTotal = sizeOfShape(tmpShape);
```
Comment (huningxin) [inline src/gather_nd.js:32]:
```suggestion
  const sliceSize = sizeOfShape(sliceShape);
```

File: src/lib/validate-input.js [hunk 6]
```diff
@@ -623,6 +623,42 @@ export function validateGatherParams(input, indices, {axis = 0} = {}) {
...
 }
 
+export function validateGatherNDParams(input, indices) {
+  const inputRank = input.rank;
+  const inputShape = input.shape;
+  const indicesRank = indices.rank;
+  const indicesShape = indices.shape;
+
+  // Refer to https://docs.openvino.ai/2024/documentation/openvino-ir-format/operation-sets/ +
+  //   operation-specs/movement/gather-nd-8.html
```
Comment (huningxin) [inline src/lib/validate-input.js:633]: Why not refer to WebNN spec?

File: src/lib/validate-input.js [hunk 7]
```diff
@@ -623,6 +623,42 @@ export function validateGatherParams(input, indices, {axis = 0} = {}) {
...
+
+  const indicesTotal = sizeOfShape(indicesShape);
+  for (let indicesIndex = 0; indicesIndex < indicesTotal; indicesIndex += lastIndicesSize) {
+    for (let i = 0; i < lastIndicesSize; i++) {
+      const indicesValue = indices.getValueByIndex(indicesIndex + i);
+      const maxSize = inputShape[i];
+      if (!Number.isInteger(indicesValue) ||
+        indicesValue < -maxSize ||
+        indicesValue > maxSize - 1) {
+        throw new Error(`Invalid indices value - it should be an integer in the interval ` +
```
Comment (huningxin) [inline src/lib/validate-input.js:655]:
```suggestion
        throw new Error(`Invalid indices value - it should be an integer in the range ` +
```

File: test/gather_nd_test.js [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+    const inputTensor = new Tensor(input.shape, input.data);
+    const indicesTensor = new Tensor(indices.shape, indices.data);
+    const outputTensor = gatherND(inputTensor, indicesTensor);
+    utils.checkShape(outputTensor, expected.shape);
+    utils.checkValue(outputTensor, expected.data);
+  }
+
+  it('gatherND with elements from 2D inputs by 2D indices', function() {
+    // Refer to Example 1 on https://docs.openvino.ai/2024/documentation/openvino-ir-format/ +
+    //   operation-sets/operation-specs/movement/gather-nd-8.html
```
Comment (huningxin) [inline test/gather_nd_test.js:18]: Same here and the following ones, URL in one line.

File: src/gather_nd.js (hunk 1 above)
Comment (BruceDai) [reply src/gather_nd.js]: Updated in new commits. Please take another look, thanks!

File: src/lib/validate-input.js (hunk 6 above)
Comment (BruceDai) [reply src/lib/validate-input.js]: Current WebNN spec doesn't define this gatherND operator, so I refer to this resources.

**Comment by BruceDai** [comment]
@huningxin I updated commits to address your comments, please take another look, thanks! ☕

**Review by fdwr — APPROVED** [review]
👍

**Review by huningxin — APPROVED** [review]
LGTM

**Pull request merged**